docs: contentful recipe (#87)
moxystudio/next-with-moxy: docusaurus/docs/recipes/contentful.md

---
id: contentful
title: Implementing Contentful
sidebar_label: Contentful
---

This recipe aims to guide you through the implementation of Contetful in the project and also provides custom solutions for different use-cases.  

### What is a CMS?

A CMS, or Content Management System, is a platform that helps in the creation and management of content to be consumed in a web application. In short, these services provide a database and a database management system with a layer of usability on top that makes them inclusive to users without technical expertise.

### Contentful

[Contenful](https://www.contentful.com/) is one such CMS, and the object of this recipe.

For more information about **Contentful**'s API please [refer to their documentation](https://www.contentful.com/developers/docs/).

> ## Modeling your database

Reviewer: I believe the correct terminology is schema and not database.

To populate the app with the fetched content, we will explore two possible implementations: one for bigger applications that justify the inclusion of Redux store, and those which don't.

#### 2.1. Accessing the client directly though `getInitialProps`

In Next.js apps, the `getInitialProps` function in general components typically receives a context object, which exposes many aspects of the context in which the page is rendered. Because we are building a custom app, and doing ourselves the work of running these `getInitialProps` functions, we can control what data they receive. In practice, we will be populating the context object with the information we want to propagate to our components. So, in your `App.js` file, you can check whether the page was loaded with a `cms-preview` query parameter, build your client, and share it across the app.

This function also has the advantage letting you return props to your App function, making it easier to signal the App itself if you're in preview mode.

```js
// App.js
import { createClient } from 'contentful';

App.getInitialProps = async ({Component, context, router}) => {

    // Create default client

>     let client = createClient({

Reviewer: Defer the creation of the client to when we know whether we should create a client pointing to a live or preview.

        space: process.env.CONTENTFUL_SPACE_ID,
        accessToken: process.env.CONTENTFUL_TOKEN, // Delivery API Token
        host: 'cdn.contentful.com', // Delivery API host
    });

    // Check whether route has `cms-preview` query parameter

>     const isPreviewingContentful = Object.hasOwnProperty.call(router.query, 'cms-preview');

Reviewer: Where does this `cms-preview` property come from exactly? By query parameter do you mean an actual URL query parameter like `?cms-preview=true`?

Reviewer: Nevermind, I completely missed this in the documentation. 😄

#### 1. Defining the content type

Firstly define the content type that will consist on a list of key/value pairs. Add the fields necessary to fit your needs, for example `homePageTitle`.

> ℹ️ Make sure that you enable localization on every field you set in your content type.

After creating your content type populate the respective fields accordingly.

#### 2. Overriding [`@moxy/next-intl`](https://github.com/moxystudio/next-intl)

Now that you have your content type set you should update the `/intl/index.js` file to fetch the data from **Contentful**. This is done by using the [**Contentful** SDK](https://www.npmjs.com/package/contentful).

```js
locales: [
    {
        id: 'en-US',
        name: 'English',
        loadMessages: async () => {
            const content = await client.getEntries({
                content_type: 'genericTranslations',
                locale: 'en-US',
                // The parameter bellow is needed to avoid the problem described at the top of the file.
            });

>             return content.items[0].fields; melhorar isto

Reviewer: Do you still need to work on this?

Reviewer 2: actually these are my leftovers :D @dominguesgm i'm sorry i didn't remove this from here on the handoff, please remove now 🙏

```js
const result = await client.getEntries({
    content_type: 'clients',
    locale: 'en-US'
});
```

### 2. Populating the app with fetched content

To populate the app with the fetched content, we will explore two possible implementations: one for bigger applications that justify the inclusion of a Redux store, and another which doesn't.

#### 2.1. Accessing the client directly though `getInitialProps`

In Next.js apps, the `getInitialProps` function in general components typically receives a context object, which exposes many aspects of the context in which the page is rendered. Because we are building a custom app, and doing the work of running these `getInitialProps` functions ourselves, we can control what data they receive. In practice, we will be populating the context object with the information we want to propagate to our components. So, in your `App.js` file, you can check whether the page was loaded with a `cms-preview` query parameter, build your client, and share it across the app.

This function also has the advantage letting you return props to your App function, making it easier to signal the App itself if you're in preview mode.

```js

> // App.js

Reviewer 3: This path should be the full like src/app/App.js. All other code blocks suffer from the same problem.

    // Create client with correct host
    const client = createClient({
        space: process.env.CONTENTFUL_SPACE_ID,
        accessToken: process.env.CONTENTFUL_TOKEN, // Delivery API Token
        host: contentfulHost, // Delivery API host
    });
    
    // Append it to the context object so other Components can access it
    context.contentfulClient = client;

    let pageProps = {}

    if (Component.getInitialProps) {
        pageProps = await Component.getInitialProps(context);
    }
    
    // Return the state here to receive as a prop,
    // allowing you to conditionally render a DOM element in case of preview
    return { pageProps, isPreviewingContentful }
};
```

Afterwards, you can use the **Contentful** client in your component's `getInitialProps` to fetch content required for that component.

```js
// UseCases.js

UseCases.getInitialProps = async (context) => {

    const { contentfulClient } = context;

    // Fetch data from Contentful
    const entries = await contentfulClient.getEntries({
        content_type: 'useCase',
    });

    // Return result as props

>     return { entries }

Reviewer 3: Missing ;

As part of using this package, you'll have to create a function (which we'll call `buildStore` ) which should return the instance of the Redux store in your application, and it will be here that you will create your `client` instance, and decide whether it should fetch published or preview content. 

Given the nature of Next.js and the implementation of [next-redux-wrapper](https://github.com/kirill-konshin/next-redux-wrapper), `buildStore` will run twice, once server-side and once client-side, requiring you to consistently instantiate the `client` to be the same for both cases. But the same strategy cannot be used for both. Each requires its own solution. 

In the following example we will be checking the presence of a `cms-preview` query parameter to decide whether to show preview content or the standard, published content, i.e. something like `www.example.com/?cms-preview`. 

```js
// buildStore.js
import { createClient } form 'contentful';
import { applyMiddleware, createStore } from 'redux';
import thunkMiddleware from 'redux-thunk';

export const buildStore = (initialState, { query }) => {

>     

Reviewer 3: Remove this empty line.

        host: 'cdn.contentful.com', // Delivery API host
    });

    if (
        // For server-side, check wether `query` exists and has the `cms-preview` key
        (typeof query !== 'undefined' && Object.hasOwnProperty.call(query, 'cms-preview')) ||
        // For client-side, check wether `window` exist and has the `cms-preview` query parameter
        (typeof window !== 'undefined' && new URLSearchParams(window.location.search).has('cms-preview'))
    ) {
        // In a positive case, switch the client instance to access the Preview API instead		
        client = createClient({
            space: process.env.CONTENTFUL_SPACE_ID,
            accessToken: process.env.CONTENTFUL_PREVIEW_TOKEN, // Preview API Token
            host: 'preview.contentful.com', // Preview API host
        });
    }

    const reducer = {
        // Example reducer
    };
    
    // Add client as a thunk middleware
    // This will give us access to the `client` object in our action functions as an argument
    const middlewares = applyMiddleware(thunkMiddleware.withExtraArgument({ client }));

    return createStore(reducer, initialState, middlewares);

>     }

Reviewer 3: Missing ;

}

export default buildStore;
```

Then, in your actions scripts, you can use the provided client to fetch the content that you need, like so:

```js
// usecases/actions.js
// Create a fetch action that fetches data from Contentful
const fetch = () => async (dispatch, getState, { client }) => {

>     dispatch({ type: 'Fetching'})

Reviewer 3: There’s a lot of missing ; in this file. Please check all other files as well.

Reviewer 3: Also there’s a missing space.

            locale: 'es-ES',
        })

        // The incoming object will be fairly complex, and it's recommended to build
        // a more conveniently structured object here so that other scrips in the pipeline
        // can avoid handling that complexity and can instead enjoy a simpler, custom built object.
        // E.G.
        const entries = result.items
            .map((item)=> ({
                title: item.fields.title,
                description: item.fields.description,
                weight: item.fields.weight,
        })).sort((a, b) => b.weight - a.weight);

        dispatch({
            type: 'Success',
            payload: { entries },
        })
    } catch (error) {
        dispatch({
            type: 'Failure',
            payload: { error },
        });
    }
};
```

### 3. Conditionally rendering DOM elements in case of preview

Since we're on the topic, you must use a similar strategy in your `App.js` file to conditionally render a DOM element to let the user know they are in a preview state. This must happen once per instance of the App, since we will not perpetuate the `cms-preview` query parameter on route changes.

> Class component:

Reviewer 3: I would put the hook version first as our boilerplate is no longer using the class version.

Reviewer 3: Moreover, next.js even removed the custom App using class components, so I don’t know if extending is the correct way here.

Author: Should we perhaps remove the Class component approach then?

Reviewer 3: Yep I think it’s better.

```js
// App.js
class App extends Nextapp {
    state = {
        isPreviewingContentful = false,
    };

    // Using the componentWillMount() lifecycle function guarantees that this runs only once per instance of App
    componentWillMount() {
        const { router } = this.props;

        this.setState({
            isPreviewingContentful: Object.hasOwnProperty.call(router.query, 'cms-preview');
        });
    }

    render() {
        return (
            (...)
                { this.isPreviewingContentful && <ContentfulPreview> }
            (...)
        );
    }
}
```

Using hooks: 

```js
// App.js
const App = ({ Component, pageProps, rootSelector, router }) => {
    const [isPreviewingContentful, setIsPreviewingContentful] = useState(false);
    
    // Using a useEffect hook with no dependencies guaranties that it fires only once per instance of App
    useEffect(() => setIsPreviewingContentful(Object.hasOwnProperty.call(router.query, 'cms-preview')), []);

    return (
        (...)

>             { this.isPreviewingContentful && <ContentfulPreview> }

Reviewer 3: `this` Is not valid here.

Missing / in the JSX tag of ContentfulPreview. Maybe we should add a note stating that it’s a component that would render a visible warning on the site saying we are on preview mode, like a ribbon. This is valid for the dato CMS recipe as well.

```js
import httpProxy from 'http-proxy';

const proxy = httpProxy.createProxyServer({});

export default (req, res) => {
    req.url = req.url.replace('/api/cms', '');
    req.headers.host = 'cdn.contentful.com';

    proxy.web(req, res, {
        target: {
            host: 'cdn.contentful.com',
        },
    });

    if (req.method === 'GET') {
        res.setHeader('Cache-Control', 's-maxage=60');
    }
};
```

Finally, you will need to setup the Contentful client to direct it's requests to the endpoint you just configured. Please note that you only want to cache the requests when in a production environment.

```js
const clientOptions = {
    space: process.env.CONTENTFUL_SPACE_ID,
    accessToken: process.env.CONTENTFUL_TOKEN,
};

if (process.env.NODE_ENV === 'production') {
    clientOptions.host = process.env.SITE_URL;
    clientOptions.basePath = '/api/contentful';
}

const client = createClient(clientOptions);
```

**NOTES:**

> > ❗️ You have to be able to access to `process.env.NODE_ENV` from the client side as well as the server side, to proxy client side requests to Contentful in production environments.

Reviewer 3: Nifty pick: always add new line after `:`, check other places too.

Moreover I would use the warning emoji if applicable to keep consistency with other recipes.

// If it is, it obtains the url to the asset that matches the id
// Otherwise it returns the content 

const getContent = (content, assets) => {
    if (isPlainObject(content)) {
        const asset = assets.find(({ sys }) => sys.id === content.id);

        return asset.fields.file.url.slice(2);
    }

    return content;
};

// Provided with the result from the Contentful API call and the entryID for the current page content,
// This function will extract the SEO data, and complete it with the URL of any needed asset.

const parseContentfulSEO = (contentfulData, entryID) => {
    const entry = contentfulData.items.find((item) => item.sys.id === entryID);
    const seoID = entry.fields.seo.id;
    const seoEntry = entry.includes.Entry.find((entry) => entry.sys.id === seoID);
    let seoContent;

    try {
        seoContent = JSON.parse(seoEntry.fields.seo);
    } catch (parsingError) {
        console.err('Unvalid JSON in SEO content');
        console.err(parsingError);
        return;
    }

    // Convert every content: { id: ''} into a content: url, keep unchanged otherwise
    const content = {
        title: seoContent.title,

>         meta: seoContent.meta.map((entry) => mapValues(entry, (value) => getContent(value, contentfulData.includes.Asset)))

Reviewer 3: Missing coma at the end of the line.

I would pass all files through the longer, there are a lot of linting and JS errors in the code blocks.

---
id: contentful
title: Implementing Contentful
sidebar_label: Contentful
---

> This recipe aims to guide you through the implementation of Contetful in the project and also provides custom solutions for different use-cases.  

Reviewer 4: ```suggestion
This recipe aims to guide you through the implementation of Contentful in the project and also provide custom solutions for different use-cases.  
```

Author: Misspell fixed, but I think `provides` is more correct

### What is a CMS?

A CMS, or Content Management System, is a platform that helps in the creation and management of content to be consumed in a web application. In short, these services provide a database and a database management system with a layer of usability on top that makes them inclusive to users without technical expertise.

### Contentful

[Contenful](https://www.contentful.com/) is one such CMS, and the object of this recipe.

For more information about **Contentful**'s API please [refer to their documentation](https://www.contentful.com/developers/docs/).

## Modeling the schema

As a developer working on the App that will consume the content stored in the CMS, you're also responsible for creating and modeling the schema itself. This will mean creating new content models that must be usable by the client team which will eventually handle **Contentful** after the hand off of the project. One **major** consideration to have is the usability and readability of the content models you're creating. Remember that they are meant to be usable without any previous technical knowledge of this context. 

> **Contentful** offers many options for validations and input appearance and descriptions to help the users understand what kind of input is expected, and you should use all of them to minimize the chance of users not understanding what's expected of them, in what ways they are limited, reducing the chances of unexpected content in your application.

Reviewer 5: ```suggestion
**Contentful** offers many options for type validations, input appearance and field descriptions to help the users understand what kind of input is expected. You should use all of them to minimize the chance of users not understanding what's expected of them, in what ways they are limited, reducing the chances of unexpected content in your application.
```

Keep in mind:

→ Always use appropriate types when possible,

> → Always use validations where appropriate,

Reviewer 5: ```suggestion
→ Use validations where appropriate to make sure the input values are what the developer expects,
```

> → Write titles and descriptions to ensure the user knows what's expected of that field.

Reviewer 5: ```suggestion
→ Use meaningful  titles and descriptions to ensure the user knows what's expected of that field.
```

## Localization

**Contentful** makes it easy to add localized content to your application. Once you define a locale, which you can do in the Locales option in the Settings menu, you can specify in each field you create whether they can be localized, in what languages, etc., giving you fine control over what needs to be localized and what doesn't. 

Once having these locales, you can then use the **Contentful** API signaling your desired locale through the `locale` field. You'll find an example of this in the next section.

## Preview

> Contentful allows you to fetch content that is not in a `published` state, also including content that is in a `changed` or `draft` state.  This can help users see how content would look on their website before publishing it and making it available on for every visitor.

Reviewer 5: ```suggestion
Contentful splits content changes in 3 states:
- `Published` : when a user saves some content.
- `Changed`: when the user changes previously saved content but still hasn't saved the new one.
- `Draft`: when the user creates new content but doesn't save.

Usually, only `Published` content is available, unless a specific request is done. This request is called `Preview` and may help users see beforehand how content would look like on their website before publishing it and making it available on for every visitor.
```

However, Contentful **does not validate content that is not published.** This means that, if your app relies on validation from Contentful, **it will likely break in preview mode unless all content in Contentful already passes validation.** This is assumed and must taken into account when developing your applications and communicated to potential users.

## Walk-through

### 1. Installing Contentful SDK

Using Contenful in your application can be done using [the **Contentful** SDK available as an npm package](https://www.npmjs.com/package/contentful), which you can install using the following command:

```sh
npm install --save contenful
```

> This SDK provides access to **Contentful**'s [Delivery AP](https://www.contentful.com/developers/docs/references/content-delivery-api/)I and [Preview API](https://www.contentful.com/developers/docs/references/content-preview-api/), both of which will be used in the integration of **Contentful** in your app. 

Reviewer 5: ```suggestion
This SDK provides access to **Contentful**'s [Delivery API](https://www.contentful.com/developers/docs/references/content-delivery-api/) and [Preview API](https://www.contentful.com/developers/docs/references/content-preview-api/), both of which will be used in the integration of **Contentful** in your app. 
```

> First, you'll need **Contentful**'s access tokens, of which you'll need three. These can all be found in the Settings menu, under the `API keys` option:

Reviewer 5: ```suggestion
First, you'll need two **Contentful**'s access tokens and a space identifier. All of the following can be found in the Settings menu, under the `API keys` option:
```

#### 1. Defining the content type

Firstly, define the content type that will consist on a list of key/value pairs. Add the fields necessary to fit your needs, for example `homePageTitle`.

> ℹ️ Make sure that you enable localization on every field you set in your content type.

After creating your content type populate the respective fields accordingly.

#### 2. Overriding [`@moxy/next-intl`](https://github.com/moxystudio/next-intl)

Now that you have your content type set you should update the `/intl/index.js` file to fetch the data from **Contentful**. This is done by using the [**Contentful** SDK](https://www.npmjs.com/package/contentful).

```js
locales: [
    {
        id: 'en-US',
        name: 'English',
        loadMessages: async () => {
            const content = await client.getEntries({
                content_type: 'genericTranslations',
                locale: 'en-US',
                // The parameter bellow is needed to avoid the problem described at the top of the file.
            });

            return content.items[0].fields;
        },
    },
],
```

## Cache

### 1. Contentful API rate limits

> API Rate limits specify the number of requests a client can make to Contentful APIs in a specific time frame. Every request counts against a per second rate limit.

Reviewer 5: ```suggestion
API Rate limits specify the maximum number of requests a client can make to Contentful APIs in a specific time frame. Every request counts against a per second rate limit.
```

    accessToken: process.env.CONTENTFUL_TOKEN,
};

if (process.env.NODE_ENV === 'production') {
    clientOptions.host = process.env.SITE_URL;
    clientOptions.basePath = '/api/contentful';
}

const client = createClient(clientOptions);
```

**NOTES:**
> ❗️ You have to be able to access to `process.env.NODE_ENV` from the client side as well as the server side, to proxy client side requests to Contentful in production environments.

> ❗️ The `process.env.SITE_URL` variable has to be correctly configured and accessible from both server and client side, otherwise the request to the proxy endpoint will not happen correctly. This may mean that preview url's in merge request will not function correctly, which will happen if the application is started with `process.env.NODE_ENV` set to `production`.

## Custom SEO

There may be cases where you will want to configure custom SEO per page. Unfortunately, Contentful does not provide out-of-the-box SEO support, so you will need to implement your own strategy. The approach we suggest is the following:

- Create a content model for SEO.
- Add a field `Title` that should be only used to identify the model, e.g. "Homepage SEO".

> - Add a SEO field (not-rich long text) to the model. Here, the SEO related tags (title, meta, etc.) will be defined in a _json_ format like so:

Reviewer 5: ```suggestion
- Add a SEO field (json type) to the model. Here, the SEO related tags (title, meta, etc.) will be defined in a _json_ format like so:
```

Author: I understand the motivation behind the suggesting, and "not-rich long text" should be changed. But the reasoning behind this was to indicate the type of input (Contentful-wise) you have to create.

Perhaps just leaving as "long text"?

Reviewer 5: ![imagem](https://user-images.githubusercontent.com/11398440/79763980-ae9bf200-831c-11ea-8fb1-1fdff3c7ec28.png)

Contentful-wise is `JSON object` :+1:

Author: 🤦 didn't see it

---
id: contentful
title: Implementing Contentful
sidebar_label: Contentful
---

This recipe aims to guide you through the implementation of Contetful in the project and also provides custom solutions for different use-cases.  

### What is a CMS?

A CMS, or Content Management System, is a platform that helps in the creation and management of content to be consumed in a web application. In short, these services provide a database and a database management system with a layer of usability on top that makes them inclusive to users without technical expertise.

### Contentful

[Contenful](https://www.contentful.com/) is one such CMS, and the object of this recipe.

For more information about **Contentful**'s API please [refer to their documentation](https://www.contentful.com/developers/docs/).

## Modeling your schema

> As a developer working on the App that will consume the content stored in the CMS, you're also responsible for creating and modeling the database itself. This will mean creating new content models that must be usable by the client team which will eventually handle **Contentful** after the hand off of the project. One **major** consideration to have is the usability and readability of the content models you're creating. Remember that they are meant to be usable without any previous technical knowledge of this context. 

Reviewer 6: > modeling the database

Change to "modeling the schema"?

Author: I don't have a personal preference, so sure, I'll change it

Reviewer: Generally, considering this is a recipe for Contentful, I'd say we should adhere to Contentful's where relevant. I've suggested "schema" instead of "database", but is "schema" the term adopted by Contentful?

Reviewer 6: Looks like Contentful calls it [data model](https://www.contentful.com/developers/docs/concepts/data-model/). But I think schema is understandable from a developer's point of view.

**Contentful** offers many options for validations and input appearance and descriptions to help the users understand what kind of input is expected, and you should use all of them to minimize the chance of users not understanding what's expected of them, in what ways they are limited, reducing the chances of unexpected content in your application.

Keep in mind:

→ Always use appropriate types when possible,

> → Always use validation where appropriate,

Reviewer 6: ```suggestion
→ Always use validations where appropriate,
```

```js
locales: [
    {
        id: 'en-US',
        name: 'English',
        loadMessages: async () => {
            const content = await client.getEntries({
                content_type: 'genericTranslations',
                locale: 'en-US',
                // The parameter bellow is needed to avoid the problem described at the top of the file.
            });

            return content.items[0].fields;
        },
    },
],
```

## Cache

### 1. Contentful API rate limits

API Rate limits specify the number of requests a client can make to Contentful APIs in a specific time frame. Every request counts against a per second rate limit.

Currently Conentful doesn't enforce any limits on requests that hit their CDN cache. For requests that do hit the Content Delivery API enforces rate limits of 78 requests per second.

When a client gets rate limited, the API responds with the `429 Too Many Requests` status code and sets the `X-Contentful-RateLimit-Reset` header that tells the client when it can make its next request. 

### 2. Custom Caching Layer

> One preventive measure for avoiding hitting the rate limit for Contentful is to implement your own custom caching layer.

Reviewer 6: > One preventive measure for avoiding hitting the rate

This sentence doesn't sound good. I would change to "One preventive measure to avoid hitting the rate".

In order to obtain the `id` used for meta tags whose content is an asset, you need to select the asset entry and, on the right side panel of the asset, select **Info** and copy the **id**. Later in the application you'll use this **id** to generate the URL for the asset.

On the application itself, the steps to customize the SEO are the following:

- Define a _default_ SEO data.
- Fetch page data as you normally would.
- Obtain the SEO data from the Contentful page data (example code for the Contentful SEO _parser_)

```js
import { mapValues } from 'lodash';

// This function will determine if a meta entry's content is an id
// If it is, it obtains the url to the asset that matches the id
// Otherwise it returns the content 

const getContent = (content, assets) => {
    if (isPlainObject(content)) {
        const asset = assets.find(({ sys }) => sys.id === content.id);

        return asset.fields.file.url.slice(2);
    }

    return content;
};

// Provided with the result from the Contentful API call and the entryID for the current page content,
// This function will extract the SEO data, and complete it with the URL of any needed asset.

> const contentfulSEOParser = (contentfulData, entryID) => {

Reviewer 6: Can we name it `parseContentfulSEO` or something similar?

Author: Yeah, I think it's a good idea to have it similar in both recipes 👍

    const entry = contentfulData.items.find((item) => item.sys.id === entryID);
    const seoID = entry.fields.seo.id;
    const seoEntry = entry.includes.Entry.find((entry) => entry.sys.id === seoID);
    let seoContent;

    try {
        seoContent = JSON.parse(seoEntry.fields.seo);
    } catch (parsingError) {
        console.err('Unvalid JSON in SEO content');
        console.err(parsingError);
        return;
    }

    // Convert every content: { id: ''} into a content: url, keep unchanged otherwise
    const content = {
        title: seoContent.title,
        meta: seoContent.meta.map((entry) => mapValues(entry, (value) => getContent(value, contentfulData.includes.Asset)))
        link: seoContent.link.map((entry) => mapValues(entry, (value) => getContent(value, contentfulData.includes.Asset)))
    }

    return content;
}
```

> - Merge the _default_ SEO data with the SEO data fetched from the Contentful.

Reviewer 6: Is this necessary? The `Seo` component already does this.

Author: The `SEO` component only accepts one object with SEO data. If you have local SEO data and SEO data from Contentful, you need to combine it in the object you will provide to the `SEO` component.

One could argue the defaults could be written directly to `Head` in the `App` component, right before the `SEO` component is rendered, but this way you wouldn't be taking advantage of the tag de-duplication behavior that the `SEO` component provides.

Reviewer 5: The SEO component uses the `Head` component internally. If we use it with the defaults in the `App` component, only the new ones set by the page will be overridden. The previous default ones that are not overridden will still be available. So there is no need to do this I think 🤔

```js
const seoData = {
    title: contentfulSEO.title || defaultSEO.title,
    meta: [...defaultSEO.meta, ...contentfulSEO.meta],
    link: [...defaultSEO.link, ...contentfulSEO.link],
}
```

- Render the SEO data (title and tags) inside the `Head` in the `App.js` file.

> For the rendering step, we recommend using [@moxy/next-seo](https://www.npmjs.com/package/@moxy/next-seo), as this package takes care of two things:

Reviewer 6: I don't think this should be only a recommendation as we already use the package in the boilerplate, I would enforce the use of `next-seo`. Also the point of creating the parser is exactly to transform the information to be used in the `Seo` component. Thoughts?

Author: Makes sense, we can change the wording from "recommendation" to "use `next-seo`".

Keep in mind:

→ Always use appropriate types when possible,

→ Use validations where appropriate to make sure the input values are what the developer expects,

→ Use meaningful titles and descriptions to ensure the user knows what's expected of that field.

## Localization

**Contentful** makes it easy to add localized content to your application. Once you define a locale, which you can do in the Locales option in the Settings menu, you can specify in each field you create whether they can be localized, in what languages, etc., giving you fine control over what needs to be localized and what doesn't. 

Once having these locales, you can then use the **Contentful** API signaling your desired locale through the `locale` field. You'll find an example of this in the next section.

## Preview

Contentful splits content changes in 3 states:

- `Published` : when a user saves some content.
- `Changed`: when the user changes previously saved content but still hasn't saved the new one.
- `Draft`: when the user creates new content but doesn't save.
  

> Usually, only `Published` content is available, unless a specific request is done. This request is called `Preview` and may help users see beforehand how content would look like on their website before publishing it and making it available on for every visitor.

Reviewer: ```suggestion
Usually, only `Published` content is available, unless a specific request is done. This request is called `Preview` and may help users see beforehand how content would look like on their website before publishing it and making it available for every visitor.
```

First, you'll need two **Contentful** access tokens and a space identifier. All three can be found in the Settings menu, under the `API keys` option:

- Space ID,
- Delivery token,
- Preview token;

Each token must be used together with their specific host, as follows:

- Delivery host: `cdn.contentful.com`,
- Preview host: `preview.contentful.com`;

Setting up the SDK can be done as follows: 

```js
import { createClient } from 'contentful';

const client = createClient({
    space: // Space ID token
    accessToken: // Either the Delivery token, or the Preview token
    host: // Either the Delivery host, or the Preview host
});
```

This `client` exposes the functions necessary to fetch the data in **Contentful**, for example:

```js
const result = await client.getEntries({ 
    content_type: 'clients',
});
```

If you're taking advantage of localization in your **Contentful** space, you must add to the request a field with your intended language code, like so:

```js
const result = await client.getEntries({
    content_type: 'clients',
    locale: 'en-US'
});
```

### 2. Populating the app with fetched content

> To populate the app with the fetched content, we will explore two possible implementations: one for bigger applications that justify the inclusion of Redux store, and those which don't.

Reviewer: Maybe I'm missing something, but the English sounded a bit strange to me. 🤔 

```suggestion
To populate the app with the fetched content, we will explore two possible implementations: one for bigger applications that justify the inclusion of a Redux store and other which doesn't.
```

Since we're on the topic, you must use a similar strategy in your `App.js` file to conditionally render a DOM element to let the user know they are in a preview state. This must happen once per instance of the App, since we will not perpetuate the `cms-preview` query parameter on route changes.

Class component:

```js
// App.js
class App extends Nextapp {
    state = {
        isPreviewingContentful = false,
    };

    // Using the componentWillMount() lifecycle function guarantees that this runs only once per instance of App
    componentWillMount() {
        const { router } = this.props;

        this.setState({
            isPreviewingContentful: Object.hasOwnProperty.call(router.query, 'cms-preview');
        });
    }

    render() {
        return (
            (...)
                { this.isPreviewingContentful && <ContentfulPreview> }
            (...)
        );
    }
}
```

Using hooks: 

```js
// App.js
const App = ({ Component, pageProps, rootSelector, router }) => {
    const [isPreviewingContentful, setIsPreviewingContentful] = useState(false);
    
    // Using a useEffect hook with no dependencies guaranties that it fires only once per instance of App

>     useEffect(() => setIsPreviewingContentful(Object.hasOwnProperty.call(router.query, 'cms-preview')), [])

Reviewer: ```suggestion
    useEffect(() => setIsPreviewingContentful(Object.hasOwnProperty.call(router.query, 'cms-preview')), [router, setIsPreviewingContentful])
```

Reviewer 7: @acostalima, as mentioned in the comment, this `useEffect` must have no dependencies so that it doesn't memoize any object that can change, this in effect guarantees that it only fires once.

This does break one of our current eslint rules though.

Also, huge thanks @dominguesgm for working on my mistakes 🙏

Reviewer: Which comment? Sorry, I'm kind of lost. 🙏

Reviewer: We can ignore `setIsPreviewingContentful` since its reference never changes anyway.

Reviewer: Can `router` change after App is mounted?

Reviewer 7: @acostalima, you are correct, we can set `router` and `setIsrPreviewingContentful`  as dependencies. My original intent was for developers not to memoize `router.query` specifically, which auto-fix in `eslint` does automatically (i.e., if you leave the dependency array empty, and run auto-fix, `eslint` will set `router.query` as a dependency).

I think the documentation should reflect this, but i'm willing to do a PR for this after this first merge if more discussion is warranted.

Reviewer: I've never came across a situation where I was using an `useEffect` referencing props or state without specifying its deps. In this case, as a general rule, we should always specify the deps as best practices suggest.

> ℹ️ Make sure that you enable localization on every field you set in your content type.

After creating your content type populate the respective fields accordingly.

#### 2. Overriding [`@moxy/next-intl`](https://github.com/moxystudio/next-intl)

Now that you have your content type set you should update the `/intl/index.js` file to fetch the data from **Contentful**. This is done by using the [**Contentful** SDK](https://www.npmjs.com/package/contentful).

```js
locales: [
    {
        id: 'en-US',
        name: 'English',
        loadMessages: async () => {
            const content = await client.getEntries({
                content_type: 'genericTranslations',
                locale: 'en-US',
                // The parameter bellow is needed to avoid the problem described at the top of the file.
            });

            return content.items[0].fields;
        },
    },
],
```

## Cache

### 1. Contentful API rate limits

API Rate limits specify the maximum number of requests a client can make to Contentful APIs in a specific time frame. Every request counts against a per second rate limit.

> Currently Conentful doesn't enforce any limits on requests that hit their CDN cache. For requests that do hit the Content Delivery API enforces rate limits of 78 requests per second.

Reviewer: ```suggestion
Currently, Contentful doesn't enforce any limits on requests that hit their CDN cache. For requests that do hit, the Content Delivery API enforces rate limits of 78 requests per second.
```
